gateway-sp-comms/src/recv_handler/mod.rs

```diff
@@ -207,22 +309,196 @@ impl RecvHandler {
             packet.data.as_ptr() as usize as u64,
             u64::from(packet.len)
         ));
-        debug!(
-            &self.log,
-            "received serial console data from {:?}: {:?}", port, packet
-        );
-        self.sp_state(port)
-            .serial_console
-            .lock()
-            .unwrap()
-            .push(packet, &self.log);
+
+        let tasks = self.sp_state(port).serial_console_tasks.lock().unwrap();
+
+        // if we have a task managing an active client connection, forward this
+        // packet to it; otherwise, drop it on the floor.
+        match tasks.get(&packet.component).map(|task| {
+            let data = packet.data[..usize::from(packet.len)].to_vec();
+            task.recv_data_from_sp(data)
+        }) {
+            Some(Ok(())) => {
+                debug!(
+                    self.log,
+                    "forwarded serial console packet to attached client"
+                );
+            }
+
+            // The only error we can get from `recv_data_from_sp()` is "the task
+            // went away", which from our point of view is the same as it not
+            // existing in the first place.
+            Some(Err(_)) | None => {
+                debug!(
+                    self.log,
+                    "discarding serial console packet (no attached client)"
+                );
+            }
+        }
     }
 }
 
 #[derive(Debug, Default)]
 struct SingleSpState {
     requests: RequestResponseMap<u32, Result<ResponseKind, ResponseError>>,
-    serial_console: Mutex<SerialConsoleHistory>,
+    serial_console_tasks: Mutex<HashMap<SpComponent, SerialConsoleTaskHandle>>,
+}
+
+#[derive(Debug)]
+struct SerialConsoleTaskHandle {
+    packets_from_sp: mpsc::UnboundedSender<Vec<u8>>,
+    detach: oneshot::Sender<()>,
+}
+
+impl SerialConsoleTaskHandle {
+    fn is_attached(&self) -> bool {
+        !self.detach.is_closed()
+    }
+    fn recv_data_from_sp(
+        &self,
+        data: Vec<u8>,
+    ) -> Result<(), mpsc::error::SendError<Vec<u8>>> {
+        self.packets_from_sp.send(data)
+    }
+}
+
+struct SerialConsoleTask {
+    communicator: Arc<Communicator>,
+    port: SwitchPort,
+    component: SpComponent,
+    detach: oneshot::Receiver<()>,
+    packets: mpsc::UnboundedReceiver<Vec<u8>>,
+    ws_stream: WebSocketStream<Upgraded>,
+    sp_ack_timeout: Duration,
+}
+
+impl SerialConsoleTask {
+    async fn run(mut self, log: &Logger) -> Result<(), SerialTaskError> {
+        use tokio_tungstenite::tungstenite::protocol::frame::coding::CloseCode;
+        use tokio_tungstenite::tungstenite::protocol::CloseFrame;
+        use tokio_tungstenite::tungstenite::Message;
+
+        let (mut ws_sink, mut ws_stream) = self.ws_stream.split();
+
+        // TODO Currently we do not apply any backpressure on the SP and are
```

Comment:
Wouldn't this then require (potentially) more memory consumption by the SP?

Reply:
Unless it was able to push the backpressure off onto the host - but I don't think that's easy to do with a UART. At least, the way the SP UART FIFO works (to the best of my knowledge) is that if the SP doesn't recv from it fast enough, it discards data until the SP catches up.

I'm not sure what the worst case failure mode is here. A websocket connection is established, and then the websocket client stays connected indefinitely but never reads? Then the server is stuck buffering up ever-growing data from the SP that it doesn't send. Leaning toward "MGS should have a max buffer size and itself discard data if the websocket gets backed up" - thoughts?